### PR TITLE
Fix how Dynamic State is tracked

### PR DESCRIPTION
**Pull request by spencer-lunarg**
Closes https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/5551 (maybe even https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/5677)

Refactor how we track dynamic state, the previous `status` had bugs, the way  around it was simply to us the same `CBDynamicFlags` in both the Command Buffer and Pipeline. Before, we were trying to track Pipeline state in `CMB_BUFFER_STATE`

### CODE REVIEW

File: layers/state_tracker/state_tracker.cpp [hunk 1]
```diff
@@ -2729,18 +2729,14 @@ void ValidationStateTracker::PreCallRecordCmdBindPipeline(VkCommandBuffer comman
         const auto *raster_state = pipe_state->RasterizationState();
         const bool rasterization_enabled = raster_state && !raster_state->rasterizerDiscardEnable;
         const auto *viewport_state = pipe_state->ViewportState();
-        const auto *dynamic_state = pipe_state->DynamicState();
-        cb_state->status &= ~cb_state->static_status;
-        cb_state->static_status = MakeStaticStateMask(dynamic_state ? dynamic_state->ptr() : nullptr);
```
Comment (spencer-lunarg) [inline layers/state_tracker/state_tracker.cpp:2734]: This logic was a hack to basically track the `PIPELINE_STATE` in the `CMD_BUFFER_STATE` but now just can call into `pipeline.dynamic_state` when needed

File: layers/state_tracker/cmd_buffer_state.cpp [hunk 2]
```diff
@@ -1078,7 +1070,20 @@ void CMD_BUFFER_STATE::UpdatePipelineState(CMD_TYPE cmd_type, const VkPipelineBi
...
-    if (pipe && VK_NULL_HANDLE != last_bound.pipeline_layout) {
+    if (!pipe) {
+        return;
+    }
+
+    // Update the consumed viewport/scissor count.
+    {
+        uint32_t &used = usedViewportScissorCount;
+        used = std::max(used, pipelineStaticViewportCount);
+        used = std::max(used, pipelineStaticScissorCount);
```
Comment (juan-lunarg) [inline layers/state_tracker/cmd_buffer_state.cpp:1081]: std::max can take a initializer_list

```suggestion
        usedViewportScissorCount = std::max({usedViewportScissorCount, pipelineStaticViewportCount, pipelineStaticScissorCount});
```

File: layers/state_tracker/cmd_buffer_state.cpp [hunk 3]
```diff
@@ -1150,7 +1155,7 @@ void CMD_BUFFER_STATE::UpdatePipelineState(CMD_TYPE cmd_type, const VkPipelineBi
             }
         }
     }
-    if (pipe && pipe->vertex_input_state && !pipe->vertex_input_state->binding_descriptions.empty()) {
+    if (pipe->vertex_input_state && !pipe->vertex_input_state->binding_descriptions.empty()) {
         vertex_buffer_used = true;
     }
```
Comment (juan-lunarg) [inline layers/state_tracker/cmd_buffer_state.cpp:1160]:
```suggestion
```

File: layers/state_tracker/cmd_buffer_state.cpp [hunk 4]
```diff
@@ -1078,7 +1070,20 @@ void CMD_BUFFER_STATE::UpdatePipelineState(CMD_TYPE cmd_type, const VkPipelineBi
...
+    // Update the consumed viewport/scissor count.
+    {
+        uint32_t &used = usedViewportScissorCount;
+        used = std::max(used, pipelineStaticViewportCount);
+        used = std::max(used, pipelineStaticScissorCount);
+        usedDynamicViewportCount |= pipe->IsDynamic(VK_DYNAMIC_STATE_VIEWPORT_WITH_COUNT);
+        usedDynamicScissorCount |= pipe->IsDynamic(VK_DYNAMIC_STATE_SCISSOR_WITH_COUNT);
+    }
+
+    if (VK_NULL_HANDLE != last_bound.pipeline_layout) {
```
Comment (juan-lunarg) [inline layers/state_tracker/cmd_buffer_state.cpp:1086]:
```suggestion
    if (last_bound.pipeline_layout == VK_NULL_HANDLE) {
        return;
    }
```

File: layers/state_tracker/state_tracker.cpp [hunk 5]
```diff
@@ -2752,14 +2748,14 @@ void ValidationStateTracker::PreCallRecordCmdBindPipeline(VkCommandBuffer comman
         // I am taking the latter interpretation based on the implementation details of NVIDIA's Vulkan driver.
         if (!has_dynamic_viewport_count) {
             cb_state->trashedViewportCount = true;
-            if (rasterization_enabled && (cb_state->static_status[CB_DYNAMIC_STATE_VIEWPORT])) {
+            if (rasterization_enabled && (!pipe_state->IsDynamic(VK_DYNAMIC_STATE_VIEWPORT))) {
                 cb_state->trashedViewportMask |= (uint32_t(1) << viewport_state->viewportCount) - 1u;
                 // should become = ~uint32_t(0) if the other interpretation is correct.
```
Comment (juan-lunarg) [inline layers/state_tracker/state_tracker.cpp:2753]: Should this be an assert?

File: layers/core_checks/cc_cmd_buffer_dynamic.cpp [hunk 6]
```diff
@@ -38,154 +37,179 @@ bool CoreChecks::ValidateDynamicStateSetStatus(const CMD_BUFFER_STATE &cb_state,
...
+        (cb_state.cmd_set_status_pipeline ^ pipeline.dynamic_state) & cb_state.cmd_set_status_pipeline;
+    if (not_set_pipeline_status.any()) {
+        skip |= LogError(objlist, vuid.dynamic_state_setting_commands_02859,
+                         "%s: %s doesn't set up %s, but it calls the related dynamic state setting commands.",
+                         CommandTypeString(cmd_type), report_data->FormatHandle(pipeline.pipeline()).c_str(),
+                         DynamicStatesToString(not_set_pipeline_status).c_str());
+    }
 
+    // build the mask of what has been set in the Pipeline, but yet to be set in the Command Buffer
+    const CBDynamicFlags not_set_cb_status = (cb_state.cmd_set_status_cb ^ pipeline.dynamic_state) & pipeline.dynamic_state;
```
Comment (juan-lunarg) [inline layers/core_checks/cc_cmd_buffer_dynamic.cpp:53]: FFTI

This variable name makes the code confusing to read. Same with `not_set_pipeline_status`.

The comments make sense. The code is just hard to follow.

In general the `not_` variables make the code harder to read.

File: layers/state_tracker/cmd_buffer_state.h [hunk 7]
```diff
@@ -180,12 +180,9 @@ class CMD_BUFFER_STATE : public REFCOUNTED_NODE {
...
 
-    // Dynamic State
-    CBDynamicFlags status;          // Track status of various bindings on cmd buffer
-    CBDynamicFlags static_status;   // All state bits provided by current graphics pipeline
-                                    // rather than dynamic state
-    CBDynamicFlags dynamic_status;  // dynamic state set up in pipeline
-
+    // Track status of all vkCmdSet* calls, if 1, means it was set
+    CBDynamicFlags cmd_set_status_cb;        // for lifetime of CommandBuffer
+    CBDynamicFlags cmd_set_status_pipeline;  // for lifetime since last bound pipeline
```
Comment (juan-lunarg) [inline layers/state_tracker/cmd_buffer_state.h:185]: FFTI

What about something like this? Having them together in a struct IMO helps readability.

```suggestion
    struct Status {
        CBDynamicFlags cb;
        CBDynamicFlags pipeline;
    } set_status;
```

File: layers/state_tracker/state_tracker.cpp (hunk 5 above)
Comment (spencer-lunarg) [reply layers/state_tracker/state_tracker.cpp:2753]: The Viewport/Scissor logic needs refactoring, but for this PR, just trying to "not break it" and going to followup after and unify this all

File: layers/state_tracker/cmd_buffer_state.h (hunk 7 above)
Comment (spencer-lunarg) [reply layers/state_tracker/cmd_buffer_state.h:185]: :clap: agree, like this

File: layers/state_tracker/cmd_buffer_state.cpp (hunk 2 above)
Comment (spencer-lunarg) [reply layers/state_tracker/cmd_buffer_state.cpp:1081]: thanks, I just moved this, but like fixing it while here

File: layers/core_checks/cc_cmd_buffer_dynamic.cpp (hunk 6 above)
Comment (spencer-lunarg) [reply layers/core_checks/cc_cmd_buffer_dynamic.cpp:53]: agree

```patch
- not_set_cb_status
+ unset_status_cb
```

Comment (artem-lunarg) [reply layers/core_checks/cc_cmd_buffer_dynamic.cpp]: Personally for me, `unset_status_cb` still have semantics of inversion but is masked by another word. Sometimes the advice is to invert the semantics, so it will end up with `set_cb_status` or just `cb_status`, not sure how it can be done here. Also, sometimes inversed term can be more natural to the problem area. Unfortunately don't have ideas about other options.

File: layers/core_checks/cc_cmd_buffer_dynamic.cpp [hunk 8]
```diff
@@ -39,154 +38,179 @@ bool CoreChecks::ValidateDynamicStateSetStatus(const LAST_BOUND_STATE &last_boun
...
-            ValidateCBDynamicStatus(cb_state, CB_DYNAMIC_STATE_DEPTH_WRITE_ENABLE, cmd_type, vuid.dynamic_depth_write_enable_07844);
-        skip |= ValidateCBDynamicStatus(cb_state, CB_DYNAMIC_STATE_DEPTH_COMPARE_OP, cmd_type, vuid.dynamic_depth_compare_op_07845);
-        skip |= ValidateCBDynamicStatus(cb_state, CB_DYNAMIC_STATE_DEPTH_BOUNDS_TEST_ENABLE, cmd_type,
-                                        vuid.dynamic_depth_bound_test_enable_07846);
-        skip |= ValidateCBDynamicStatus(cb_state, CB_DYNAMIC_STATE_STENCIL_TEST_ENABLE, cmd_type,
-                                        vuid.dynamic_stencil_test_enable_07847);
-        skip |= ValidateCBDynamicStatus(cb_state, CB_DYNAMIC_STATE_STENCIL_OP, cmd_type, vuid.dynamic_stencil_op_07848);
-        skip |= ValidateCBDynamicStatus(cb_state, CB_DYNAMIC_STATE_VERTEX_INPUT_BINDING_STRIDE, cmd_type,
-                                        vuid.vertex_input_binding_stride_04884);
+        skip |= ValidateDynamicStateIsSet(not_set_cb_status, CB_DYNAMIC_STATE_CULL_MODE, objlist, cmd_type,
```
Comment (artem-lunarg) [inline layers/core_checks/cc_cmd_buffer_dynamic.cpp]: Question. Each `ValidateDynamicStateIsSet` call validations a single flag. Is it allowed to have some flag set in `not_set_cb_status` that according to logic should not be checked, or does `ValidateDynamicStateIsSet` have to check all possible flags?

File: layers/state_tracker/cmd_buffer_state.cpp [hunk 9]
```diff
@@ -1078,7 +1070,21 @@ void CMD_BUFFER_STATE::UpdatePipelineState(CMD_TYPE cmd_type, const VkPipelineBi
...
+    }
+
+    // Update the consumed viewport/scissor count.
+    {
+        usedViewportScissorCount = std::max({usedViewportScissorCount, pipelineStaticViewportCount, pipelineStaticScissorCount});
+        usedDynamicViewportCount |= pipe->IsDynamic(VK_DYNAMIC_STATE_VIEWPORT_WITH_COUNT);
+        usedDynamicScissorCount |= pipe->IsDynamic(VK_DYNAMIC_STATE_SCISSOR_WITH_COUNT);
+    }
+
+    if (last_bound.pipeline_layout != VK_NULL_HANDLE) {
```
Comment (juan-lunarg) [inline layers/state_tracker/cmd_buffer_state.cpp:1087]: You can return early. This avoids double nested logic.

```suggestion
    if (last_bound.pipeline_layout == VK_NULL_HANDLE) {
        return;
    }
```

File: layers/core_checks/cc_cmd_buffer_dynamic.cpp [hunk 10]
```diff
@@ -23,12 +23,11 @@
 #include "generated/chassis.h"
 #include "core_validation.h"
 
-// Check object status for selected flag state
-bool CoreChecks::ValidateCBDynamicStatus(const CMD_BUFFER_STATE &cb_state, CBDynamicState dynamic_state, CMD_TYPE cmd_type,
-                                         const char *msg_code) const {
-    if (!(cb_state.status[dynamic_state])) {
-        return LogError(cb_state.commandBuffer(), msg_code, "%s: %s state not set for this command buffer.",
-                        CommandTypeString(cmd_type), DynamicStateToString(dynamic_state));
```
Comment (juan-lunarg) [inline layers/core_checks/cc_cmd_buffer_dynamic.cpp:31]: This is much more readable to me then the new version.

If the intent behind the `not_set_status` is to remove the `!`

What about just returning early?

```
if (cb_state.status[dynamic_state]) {
    return false;
}

return LogError(...);
```

File: layers/state_tracker/cmd_buffer_state.h [hunk 11]
```diff
@@ -216,14 +216,16 @@ class CMD_BUFFER_STATE : public REFCOUNTED_NODE {
         // VK_DYNAMIC_STATE_LINE_STIPPLE_ENABLE_EXT
         bool stippled_line_enable;
 
+        uint32_t color_write_enable_attachment_count;
```
Comment (juan-lunarg) [inline layers/state_tracker/cmd_buffer_state.h:219]: Initialize this in the header file. 


```suggestion
        uint32_t color_write_enable_attachment_count = 0;
```

Also does this value need to be set to 0 when reseting? I only see this value being assigned a value in 1 location.

IE should this value be set to 0 in the reset helper function.

File: layers/state_tracker/state_tracker.cpp [hunk 12]
```diff
@@ -5441,24 +5437,17 @@ void ValidationStateTracker::PostCallRecordCmdSetVertexInputEXT(
...
     }
     cb_state->RecordStateCmd(CMD_SETVERTEXINPUTEXT, status_flags);
 }
 
 void ValidationStateTracker::PostCallRecordCmdSetColorWriteEnableEXT(VkCommandBuffer commandBuffer, uint32_t attachmentCount,
                                                                      const VkBool32 *pColorWriteEnables) {
     auto cb_state = GetWrite<CMD_BUFFER_STATE>(commandBuffer);
-    cb_state->RecordColorWriteEnableStateCmd(CMD_SETCOLORWRITEENABLEEXT, CB_DYNAMIC_STATE_COLOR_WRITE_ENABLE_EXT, attachmentCount);
+    cb_state->RecordStateCmd(CMD_SETCOLORWRITEENABLEEXT, CB_DYNAMIC_STATE_COLOR_WRITE_ENABLE_EXT);
+    cb_state->dynamic_state_value.color_write_enable_attachment_count = attachmentCount;
```
Comment (juan-lunarg) [inline layers/state_tracker/state_tracker.cpp:5450]: This is what I'm referencing I see this value only being assigned here.

So on the next frame if the command buffer is used again and doesn't call `CmdSetColorWriteEnableEXT` what is the value of `color_write_enable_attachment_count`?

File: layers/state_tracker/cmd_buffer_state.cpp (hunk 9 above)
Comment (spencer-lunarg) [reply layers/state_tracker/cmd_buffer_state.cpp:1087]: so the issue is when we decide to add more logic under to add to `UpdatePipelineState` we now have to be careful, either make sure to add it above, or risk a bug someone didn't see this sudden `return` case in the middle of the function (I have hit this a few times before in the VVL with early returns mid-functions)

I think if we are nesting to much, it should be pulled out to another function

File: layers/state_tracker/cmd_buffer_state.h (hunk 11 above)
Comment (spencer-lunarg) [reply layers/state_tracker/cmd_buffer_state.h:219]: right below

```
 // When the Command Buffer resets, the value most things in this struct don't matter because if they are read without
 // setting the state, it will fail in ValidateDynamicStateIsSet() for us
```

File: layers/core_checks/cc_cmd_buffer_dynamic.cpp (hunk 6 above)
Comment (spencer-lunarg) [reply layers/core_checks/cc_cmd_buffer_dynamic.cpp]: ok, so `unset` was suppose to mean "Dynamic State non set", which means "static state set" so can rename this to `CBDynamicFlags static_state_cb`

File: layers/state_tracker/state_tracker.cpp (hunk 12 above)
Comment (spencer-lunarg) [reply layers/state_tracker/state_tracker.cpp:5450]: this is `dynamic_state_value.color_write_enable_attachment_count`

so only used when things are dynamic

>  and doesn't call CmdSetColorWriteEnableEXT

You hit this VU

> VUID-vkCmdDraw-None-07749
> If the bound graphics pipeline state was created with the VK_DYNAMIC_STATE_COLOR_WRITE_ENABLE_EXT dynamic state enabled then [vkCmdSetColorWriteEnableEXT](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#vkCmdSetColorWriteEnableEXT) must have been called in the current command buffer prior to this drawing command

Basically the values are "undefined" in the spec so "resetting" them to some arbitrary value like zero, means nothing

File: layers/core_checks/cc_cmd_buffer_dynamic.cpp (hunk 10 above)
Comment (spencer-lunarg) [reply layers/core_checks/cc_cmd_buffer_dynamic.cpp:31]: yep renamed, agree was not well worded

File: layers/core_checks/cc_cmd_buffer_dynamic.cpp (hunk 8 above)
Comment (spencer-lunarg) [reply layers/core_checks/cc_cmd_buffer_dynamic.cpp]: >  Is it allowed to have some flag set in not_set_cb_status that according to logic should not be checked

yes, below you will find there are some more complex checks that are made, these are the "simple" ones

File: layers/core_checks/cc_cmd_buffer_dynamic.cpp [hunk 13]
```diff
@@ -39,154 +38,180 @@ bool CoreChecks::ValidateDynamicStateSetStatus(const LAST_BOUND_STATE &last_boun
...
+        (cb_state.dynamic_state_status.pipeline ^ pipeline.dynamic_state) & cb_state.dynamic_state_status.pipeline;
+    if (unset_status_pipeline.any()) {
+        skip |= LogError(objlist, vuid.dynamic_state_setting_commands_02859,
+                         "%s: %s doesn't set up %s, but it calls the related dynamic state setting commands.",
+                         CommandTypeString(cmd_type), report_data->FormatHandle(pipeline.pipeline()).c_str(),
+                         DynamicStatesToString(unset_status_pipeline).c_str());
+    }
+
+    // build the mask of what has been set in the Pipeline, but yet to be set in the Command Buffer
+    const CBDynamicFlags state_status_cb = ~((cb_state.dynamic_state_status.cb ^ pipeline.dynamic_state) & pipeline.dynamic_state);
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_cmd_buffer_dynamic.cpp:54]: @juan-lunarg @artem-lunarg sorry, I just flipped the logic here, remove the `un`/`non` in the name, and keeps the logic easier in `ValidateDynamicStateIsSet` as well

This whole Dynamic Code will most likely need to be fix when we add `VK_EXT_shader_object`

**Review by juan-lunarg — COMMENTED** [review]
LGTM

File: layers/state_tracker/cmd_buffer_state.cpp (hunk 9 above)
Comment (juan-lunarg) [reply layers/state_tracker/cmd_buffer_state.cpp:1087]: > or risk a bug someone didn't see this sudden return case in the middle of the function

I see your point. But also that is exactly why we try to require tests when writing validation.
